### ISSUE TITLE
react/display-name: anonymous component returned by a curried HOC is missed; diagnostic anchors on the outer arrow

### ISSUE THREAD

**Issue by ahmedelgabri** [oxc-project/oxc#25618]
## Summary

`react/display-name` does not detect a missing display name on an inline anonymous component returned by a curried higher-order component. It anchors the diagnostic on the *outer* arrow (the HOC factory, which returns a function — not JSX) and misses the *inner* returned arrow, which returns JSX and is the real anonymous component. `eslint-plugin-react` on both ESLint 7 and ESLint 9 flags the inner arrow. As a result, an `// eslint-disable-next-line react/display-name` placed above the inner arrow (the ESLint-compatible location) is reported as an unused directive while the outer-anchored error still fires.

## oxlint version

1.78.0; originally observed on 1.69.0, re-verified on 1.72.0 and 1.78.0.

## Reproduction

`repro.tsx`:

```tsx
import React from 'react'

type Props = {onSubmit: () => void}
type HOCProps = {label: string}

export const HOC =
	({label}: HOCProps) =>
	({onSubmit}: Props): JSX.Element => <button onClick={onSubmit}>{label}</button>
```

`.oxlintrc.json` (enabling the native rule only):

```json
{
	"plugins": ["react"],
	"categories": {"correctness": "off"},
	"settings": {"react": {"version": "18.3"}},
	"rules": {"react/display-name": "error"}
}
```

Run:

```
oxlint -c .oxlintrc.json repro.tsx
```

For the directive variant, add `// eslint-disable-next-line react/display-name` on the line above the inner arrow (line 8) and run with `--report-unused-disable-directives`.

## Expected

ESLint (`eslint-plugin-react` 7.37.5, on both ESLint 7 and ESLint 9) reports `react/display-name` ("Component definition is missing display name") on the **inner** returned arrow — `repro.tsx:8:2`, the `({onSubmit}: Props): JSX.Element => <button .../>` line — because a function that returns JSX is treated as a component definition, and this one is anonymous. With the disable directive above the inner arrow, ESLint exits clean: the directive suppresses the violation and is not unused.

## Actual

oxlint 1.78.0 anchors the diagnostic on the **outer** arrow (the `HOC` declaration) and never flags the inner JSX-returning arrow:

```
repro.tsx:6:14: error react(display-name): Component definition is missing display name.
```

With the directive above the inner arrow:

```
repro-disable.tsx:6:14: error react(display-name): Component definition is missing display name.
repro-disable.tsx:8:2: warning: Unused eslint-disable directive (no problems were reported).
```

The directive suppresses nothing, is reported unused, and the mis-anchored error still fires.

## Cross-checked

| Tool | Result on `repro.tsx` | Directive above inner arrow |
| --- | --- | --- |
| ESLint 7.23.0 + eslint-plugin-react 7.37.5 (@typescript-eslint/parser 6.21.0) | error at 8:2 (inner arrow) | clean — directive used |
| ESLint 9.39.4 flat config + eslint-plugin-react 7.37.5 (typescript-eslint 8.62.1) | error at 8:2 (inner arrow) | clean — directive used |
| oxlint 1.78.0 | error at 6:14 (outer arrow); inner arrow missed | 6:14 error + "Unused eslint-disable directive" |
